src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SslOverTdsStream.NetCoreApp.cs

```diff
@@ -0,0 +1,289 @@
+
+                    Span<byte> data = packetBuffer.AsSpan(TdsEnums.HEADER_LEN, dataLength);
+                    remaining.Slice(0, dataLength).CopyTo(data);
+
+                    _stream.Write(packetBuffer.AsSpan(0, packetLength));
+                    _stream.Flush();
+
+                    remaining = remaining.Slice(dataLength);
+                }
+                if (packetBuffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(packetBuffer, clearArray: true);
+                }
+            }
+            else
+            {
+                _stream.Write(buffer);
+                _stream.Flush();
+            }
+        }
+
+        public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (_encapsulate)
+            {
+                ReadOnlyMemory<byte> remaining = buffer;
+                byte[] packetBuffer = null;
+                while (remaining.Length > 0)
+                {
+                    int dataLength = Math.Min(PACKET_SIZE_WITHOUT_HEADER, remaining.Length);
+                    int packetLength = TdsEnums.HEADER_LEN + dataLength;
+
+                    if (packetBuffer == null)
+                    {
+                        packetBuffer = ArrayPool<byte>.Shared.Rent(packetLength);
+                    }
+                    else if (packetBuffer.Length < packetLength)
+                    {
+                        ArrayPool<byte>.Shared.Return(packetBuffer, clearArray: true);
+                        packetBuffer = ArrayPool<byte>.Shared.Rent(packetLength);
+                    }
+
+                    SetupPreLoginPacketHeader(packetBuffer, dataLength, remaining.Length - dataLength);
+
+                    remaining.Span.Slice(0, dataLength).CopyTo(packetBuffer.AsSpan(TdsEnums.HEADER_LEN, dataLength));
+
+                    {
+                        ValueTask packetWriteValueTask = _stream.WriteAsync(new ReadOnlyMemory<byte>(packetBuffer, 0, packetLength), cancellationToken);
+                        if (!packetWriteValueTask.IsCompletedSuccessfully)
+                        {
+                            await packetWriteValueTask.AsTask().ConfigureAwait(false);
+                        }
+                    }
+
+                    await _stream.FlushAsync().ConfigureAwait(false);
+
+
+                    remaining = remaining.Slice(dataLength);
+                }
+                if (packetBuffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(packetBuffer, clearArray: true);
+                }
+            }
+            else
+            {
+                {
+                    ValueTask valueTask = _stream.WriteAsync(buffer, cancellationToken);
+                    if (!valueTask.IsCompletedSuccessfully)
+                    {
+                        await valueTask.AsTask().ConfigureAwait(false);
+                    }
+                }
```

Comment:
When using ValueTask you must be careful to get either the result or the task only once. To enforce this I isolate them into an additional scope so that if anyone tries to use them again they will get an error with them being out of scope,

Reply:
Why need an `.AsTask()` here, should it just be `await valueTask.ConfigureAwait(false);` instead. Also for other place use `await xxx.AsTask().ConfigureAwait(false)`

Reply:
@yyjdelete see below discussion in https://github.com/dotnet/SqlClient/pull/541#discussion_r415412344.

Reply:
Not tested whether it affact the perf, but `AsTask()` may need to create an new Task with the `IValueTaskSource`, if the valueTask wraps an `IValueTaskSource` instead of `Task`(NetworkStream already use `IValueTaskSource` in netcoreapp if possible)

Reply:
I don't think there's any need to have AsTask. It's perfectly fine to directly await a ValueTask.

Reply:
Should I await ValueTask or call AsTask() for the awaiter? Not sure I care really. [Sharplab shows that](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgBgAJDSBWAbipJveIGY6BvKnSENixfoOGTGATgAUAIkA8G4BB9+QEo2lSQF8JQvQ16MkDABx0AagENcAVxiEkAHmwA7DAD46AWRgYAFhAAJrJqBgJaksJuGHSwGABuNnQAvHQ0mlHC1nYOzjFeCbFpdK4wAO6WNvaOLu4esqTEGgaS2ABmdLJFAHQAkgDOAMIQALYADrh+MEEAyrZgYDADA+22uLgAnmGRWRFZWfFJuKl0vQBKy+sYmQe6u1EwuAMwrcL7B5JHyWmE0mc3T7CN5Ce6fQgAdjifmOt1BBio2iAA==) it makes a single IL instruction difference (which admittedly is a callvirt) and since I'm not going to be able to sensibly measure the perf of this in-situ I'm not going to worry about it unless a profile shows that I need to.

As noted in https://github.com/dotnet/SqlClient/pull/541#discussion_r415418591 I used the pattern I've seen used in corefx and documented in https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/  because I'm not trying to make perf gains here but there's a pattern that can be used casually to make perf good by default so I used it. @Roji noted that this only perhaps saved 7ns, which is fine with me if the library owners are ok with it, if not they'll request the simplified version.

So if you want to micro optimize the hell out of this or fun then go ahead but I'm not sure it's helpful to do it in this PR. If you're looking for stuff to improve managed performance then it'd be good if you would look over one of my other PR's which has been open since January with no movement https://github.com/dotnet/SqlClient/pull/389 . It contains interesting packet recycling mechanics and an entirely measurable 7% overall read perf increase. For general performance the removal of a lot of allocations on async paths using SqlDataReader is in https://github.com/dotnet/SqlClient/pull/528 that could use some careful review and if you're feeling like a really deep dive into the library then I opened https://github.com/dotnet/SqlClient/issues/533 to track the cause of the 2nd and 4th most allocated objects in all async reads.